src/viewport/s3_service.py

```
@@ -330,13 +377,15 @@ async def file_exists(self, key: str) -> bool:
             async with self._get_s3_client() as s3:
                 await s3.head_object(Bucket=self.settings.bucket, Key=key)
+            record_s3_operation("exists", "success")
             return True
         except ClientError as e:
-            # Check if it's a NoSuchKey error by checking the exception code
             error_code = getattr(e.response, "Error", {}).get("Code") if hasattr(e, "response") else None
             if error_code == "404" or "NoSuchKey" in str(type(e).__name__):
+                record_s3_operation("exists", "not_found")
                 return False
-            logger.error("Failed to check if object exists %s: %s", key, e)
+            logger.error("Failed to check if object exists key_hash=%s: %s", _key_hash(key), safe_exception_summary(e))
```

Ruff (0.15.12)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)